libs/traits/src/lib.rs

```diff
@@ -695,3 +695,71 @@ pub mod fees {
 		}
 	}
 }
+
+mod ops {
+	pub use sp_runtime::{
+		traits::{CheckedAdd, CheckedDiv, CheckedMul, CheckedSub},
+		ArithmeticError,
+	};
+
+	/// Performs addition that returns `ArithmeticError::Overflow` instead of
+	/// wrapping around on overflow.
+	pub trait EnsureAdd: CheckedAdd {
+		/// Adds two numbers, checking for overflow. If overflow happens,
+		/// `ArithmeticError::Overflow` is returned.
+		///
+		/// ```
+		/// u32::max().ensure_add(1)?
+		/// ```
+		fn ensure_add(&self, v: &Self) -> Result<Self, ArithmeticError> {
+			self.checked_add(v).ok_or(ArithmeticError::Overflow)
+		}
+	}
+
+	/// Performs subtraction that returns `ArithmeticError::Underflow` instead of
+	/// wrapping around on underflow.
+	pub trait EnsureSub: CheckedSub {
+		/// Subtracts two numbers, checking for overflow. If overflow happens,
+		/// `ArithmeticError::Underflow` is returned.
+		///
+		/// ```
+		/// 0u32.ensure_sub(1)?
+		/// ```
+		fn ensure_sub(&self, v: &Self) -> Result<Self, ArithmeticError> {
+			self.checked_sub(v).ok_or(ArithmeticError::Underflow)
+		}
+	}
+
+	/// Performs multiplication that returns `ArithmeticError::Overflow` instead of
+	/// wrapping around on overflow.
+	pub trait EnsureMul: CheckedMul {
+		/// Multiplies two numbers, checking for overflow. If overflow happens,
+		/// `ArithmeticError::Overflow` is returned.
+		///
+		/// ```
+		/// u32::max().ensure_mul(2)?
+		/// ```
+		fn ensure_mul(&self, v: &Self) -> Result<Self, ArithmeticError> {
+			self.checked_mul(v).ok_or(ArithmeticError::Overflow)
```

Comment:
I leave a tech debt here to fix the case where the overflow is, in fact, underflow.

Reply:
We can write a small macro and only implement it for unsigned and signed? WDYT?

Reply:
Maybe with two different macros:
- If it's unsigned, ok, it will always be Overflow. Current behavior.
- If it's signed, it can be Overflow or Underflow, depending on the sign of multiplying both numbers.

Reply:
Maybe lets make solely impls for unsigned. We have no signed logic currently

I was thinking of something like this.
```rust
ensure_ops_signed_impl!(u8);
ensure_ops_signed_impl!(u16);
ensure_ops_signed_impl!(32);
ensure_ops_signed_impl!(u64);
ensure_ops_signed_impl!(u128);
```